targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/README.md

```diff
@@ -0,0 +1,72 @@
+# Cortex-M3 Design Start Eval package example on MPS2+ board
+
+This folder includes the port of mbed OS on the example system of the Cortex-M3
+Design Start Eval package. This example is implemented in FPGA on the MPS2+
+board. Please check the [mbed page](https://os.mbed.com/platforms/ARM-CM3DS/) of
+this target for more information.
+For convenience, this target is called **CM3DS**.
+
+## Compiling :construction_worker:
+
+The target name is `ARM_CM3DS_MPS2`, you should be able to compile mbed OS
+projects for CM3DS with:
+
+```bash
+mbed compile -t COMPILER -m ARM_CM3DS_MPS2
+```
+
+The following compilers are supported (replace `COMPILER` with):
+
+* `ARM` for Arm Compiler version 5
+* `GCC_ARM` for GNU Compiler for Arm
+* `IAR` for IAR compiler
+
+## Running :running:
+
+Because of the new memory configuration introduced in commit `CM3DS: switch to
+larger memories for code and data` it
+has become more easy (and portable amoung all compilers) to use `.elf` files
+instead of `.bin`.
+For `.elf` files to work, you will need **at least version 2.2.5** of the MPS2+
+firmware. You can find firmware version 2.2.6 and instructions on how to put it
+in the MPS2+ board [here](https://community.arm.com/processors/designstart/f/discussions/9727/mps2-firmware-for-mbed).
+
+## Testing :zap:
+
+If you want to execute the mbed OS greentea tests on CM3DS, you will need
+**at least firmware version 2.2.6**.
+
+* `mbedls` does not automatically recognize which serial port is linked to the
+board. Check it manually and create a file named `mbedls.json` containing
+(at the same level than where you execute all commands):
+```bash
+{
+    "50040200074D652F3828F333": {
+        "serial_port": "/dev/ttyUSB0"
+    }
+}
+```
+Replace `/dev/ttyUSB0` with your correct serial port
+(something like `COM6` on Windows).
+
+* `mbedls` does not link CM3DS target ID with its name so execute the command:
+```bash
+mbedls --mock 5004:ARM_CM3DS_MPS2
+```
+
+* Because greentea works with `.bin` files, you will have to first compile the
+tests:
+```bash
+mbed test --compile -m ARM_CM3DS_MPS2 -t COMPILER
```

Comment:
Please provide an application configuration that  overrides the config Variable `target.OUTPUT_EXT` to `"elf"`. Then pass `--app-config` with the path to you're application configuration.

Reply:
> Good catch! Did you need to do the first part of that change? or do you prefer == over is?

Thanks. Yes, it is needed. It was the original bug, [it seems](https://stackoverflow.com/questions/132988/is-there-a-difference-between-and-is-in-python) that in Python `is` compares the address of the objects while `==` compares their content. With `is`, the `bin` variable was not `None` which triggered the `elf2bin` conversion and leading to another issue further in the code.

> It's a bug fix version. Please submit that change as another PR.

Allright! I will submit another PR.

Can I also suggest to add `"OUTPUT_EXT": "elf"` in the `targets/target.json` file part for `CM3DS` instead of giving an application configuration? After this change, `elf` files should become the "normal" binary type for `CM3DS` anyway. What do you think?

Reply:
Thanks for the explination! I was incorrectly under the impression that python interned small strings.

------

Yes, lets' have it in `targetsvtargets.json`. That'll be easiest for everyone. Bonus: You no longer need two commands for test! you should be able to `mbed test -m ... -t ...`.

Reply:
Cool, I will change that in the new patch.